In [1]:
# import libaries

import pandas as pd
import numpy as np;
import os

In [2]:
# importing data

path = r'/Users/lennartzeidler/Desktop/Uni/Instacart Basket Analysis 08.10.2022 (4)'

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))
df_orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))

In [5]:
df_orders_products_combined = df_orders_products_combined.drop(columns=['_merge'])

### Consistency checks

In [8]:
df_orders_products_combined.shape

(32434489, 9)

In [9]:
# check for mixed types
for col in df_orders_products_combined.columns.tolist():
  weird = (df_orders_products_combined[[col]].applymap(type) != df_orders_products_combined[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_orders_products_combined[weird]) > 0:
    print (col)

In [13]:
df_orders_products_combined.isnull().sum()

order_id                        0
user_id                         0
order_number                    0
order_day_of_week               0
order_hour_of_day               0
days_since_prior_order    2078068
product_id                      0
add_to_cart_order               0
reordered                       0
dtype: int64

In [15]:
df_orders_products_combined.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.243449e+07  3.243449e+07  3.243449e+07       3.243449e+07   
mean   1.710749e+06  1.029372e+05  1.714205e+01       2.738818e+00   
std    9.873007e+05  5.946648e+04  1.753504e+01       2.090049e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.559430e+05  5.142100e+04  5.000000e+00       1.000000e+00   
50%    1.711048e+06  1.026110e+05  1.100000e+01       3.000000e+00   
75%    2.565514e+06  1.543910e+05  2.400000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01       6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.243449e+07            3.035642e+07  3.243449e+07   
mean        1.342498e+01            1.110407e+01  2.557634e+04   
std         4.246365e+00            8.778914e+00  1.409669e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.353000e+04   
50%         1.300000e+01            8.000000e+00  2.525600e+04   
75%         1.600000e+01            1.500000e+01  3.793500e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered  
count       3.243449e+07  3.243449e+07  
mean        8.351076e+00  5.896975e-01  
std         7.126671e+00  4.918886e-01  
min         1.000000e+00  0.000000e+00  
25%         3.000000e+00  0.000000e+00  
50%         6.000000e+00  1.000000e+00  
75%         1.100000e+01  1.000000e+00  
max         1.450000e+02  1.000000e+00

### Merging process

In [19]:
# merging data with inner join
df_orders_products_merged = df_orders_products_combined.merge(df_prods, on = 'product_id', indicator = True)

In [22]:
# merging data with outer join for testings
df_orders_products_merged_test = df_orders_products_combined.merge(df_prods, on = 'product_id', indicator = True, how = 'outer')

In [20]:
df_orders_products_merged['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [23]:
df_orders_products_merged_test['_merge'].value_counts()

both          32404859
left_only        30200
right_only          11
Name: _merge, dtype: int64

In [25]:
# checking for null values just to be save
df_orders_products_merged.isnull().sum()

order_id                        0
user_id                         0
order_number                    0
order_day_of_week               0
order_hour_of_day               0
days_since_prior_order    2076096
product_id                      0
add_to_cart_order               0
reordered                       0
product_name                    0
aisle_id                        0
department_id                   0
prices                          0
_merge                          0
dtype: int64

#### results
We are losing a couple of rows through the inner join, however it was said that I only work with full matched data. That means I stick to the inner join with 32404859 columns.

In [26]:
# exporting data in pickle format because of the df size
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))